controllers/apps/operations/restore.go

```diff
@@ -0,0 +1,202 @@
+			opsRequest.Labels = make(map[string]string)
+		}
+		opsRequest.Labels[constant.AppInstanceLabelKey] = opsRequest.Spec.ClusterRef
+		opsRequest.Labels[constant.OpsRequestTypeLabelKey] = string(opsRequest.Spec.Type)
+		scheme, _ := appsv1alpha1.SchemeBuilder.Build()
+		if err := controllerutil.SetOwnerReference(cluster, opsRequest, scheme); err != nil {
+			return err
+		}
+		if err := cli.Patch(reqCtx.Ctx, opsRequest, patch); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ReconcileAction implements the restore action.
+// It will check the cluster status and update the OpsRequest status.
+// If the cluster is running, it will update the OpsRequest status to Complete.
+// If the cluster is failed, it will update the OpsRequest status to Failed.
+// If the cluster is not running, it will update the OpsRequest status to Running.
+func (r RestoreOpsHandler) ReconcileAction(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) (appsv1alpha1.OpsPhase, time.Duration, error) {
+	opsRequest := opsRes.OpsRequest
+	clusterDef := opsRequest.Spec.ClusterRef
+
+	// get cluster
+	cluster := &appsv1alpha1.Cluster{}
+	if err := cli.Get(reqCtx.Ctx, client.ObjectKey{
+		Namespace: opsRequest.GetNamespace(),
+		Name:      clusterDef,
+	}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			_ = PatchClusterNotFound(reqCtx.Ctx, cli, opsRes)
+		}
+		return appsv1alpha1.OpsFailedPhase, 0, err
+	}
+
+	// check if the cluster is running
+	if cluster.Status.Phase == appsv1alpha1.RunningClusterPhase {
+		return appsv1alpha1.OpsSucceedPhase, 0, nil
+	} else if cluster.Status.Phase == appsv1alpha1.FailedClusterPhase {
+		return appsv1alpha1.OpsFailedPhase, 0, fmt.Errorf("restore failed")
+	}
+	return appsv1alpha1.OpsRunningPhase, 0, nil
+}
+
+// SaveLastConfiguration saves last configuration to the OpsRequest.status.lastConfiguration
+func (r RestoreOpsHandler) SaveLastConfiguration(reqCtx intctrlutil.RequestCtx, cli client.Client, opsResource *OpsResource) error {
+	return nil
+}
+
+func restoreClusterFromBackup(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRequest *appsv1alpha1.OpsRequest, clusterDef string) (*appsv1alpha1.Cluster, error) {
+	backupName := opsRequest.Spec.RestoreSpec.BackupName
+	restoreTimeStr := opsRequest.Spec.RestoreSpec.RestoreTimeStr
+	volumeRestorePolicy := opsRequest.Spec.RestoreSpec.VolumeRestorePolicy
+
+	// check if the backup exists
+	backup := &dpv1alpha1.Backup{}
+	if err := cli.Get(reqCtx.Ctx, client.ObjectKey{
+		Name:      backupName,
+		Namespace: opsRequest.Namespace,
+	}, backup); err != nil {
+		return nil, err
+	}
+
+	// check if the backup whether is completed
+	if backup.Status.Phase != dpv1alpha1.BackupPhaseCompleted {
+		return nil, fmt.Errorf("backup %s is not completed", backupName)
+	}
+	if len(backup.Labels[constant.AppInstanceLabelKey]) == 0 {
+		return nil, fmt.Errorf(`missing source cluster in backup "%s", "app.kubernetes.io/instance" is empty in labels`, backupName)
+	}
```

Comment:
why need  source cluster?

Reply:
I refer to the restore function in kbcli. Firstly try to get cluster from annotation, if annotation is empty, then get cluster from source cluster. But I don't know if it's necessary. Maybe getting the cluster from annotation is ok.
```go
// use the cluster snapshot to restore firstly
	clusterString, ok := backup.Annotations[constant.ClusterSnapshotAnnotationKey]
	if ok {
		if err := json.Unmarshal([]byte(clusterString), &cluster); err != nil {
			return nil, err
		}
	} else {
		clusterName := backup.Labels[constant.AppInstanceLabelKey]
		if err := cli.Get(context.Background(), client.ObjectKey{
			Namespace: backup.Namespace,
			Name:      clusterName,
		}, cluster); err != nil {
			return nil, err
		}
	}
```

Reply:
I think, the source cluster is not necessary, only depend on the cluster snapshot in the annotation.